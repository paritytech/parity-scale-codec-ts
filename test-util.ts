/// <reference lib="deno.unstable"/>

import { dirname, join } from "std/path/win32.ts";
import { assertEquals } from "std/testing/asserts.ts";
import { assertSnapshot } from "std/testing/snapshot.ts";
import { Codec } from "./common.ts";

const file = (path: string) => () => fetch(join(dirname(import.meta.url), path)).then((x) => x.text());
export const files = {
  lipsum: file("lipsum.txt"),
  words: file("words.txt"),
  cargoLock: file("Cargo.lock"),
};

export function testCodec<T>(name: string, codec: Codec<T>, values: T[] | Record<string, T>) {
  for (const key in values) {
    const value = values[key as never] as T;
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key;
    Deno.test(`${name} ${label}`, async (t) => {
      const encoded = codec.encode(value);
      const decoded = codec.decode(encoded);
      assertEquals(decoded, value);
      await assertSnapshot(t, encoded, { serializer: serializeU8A });
    });
  }
}

function serializeU8A(array: Uint8Array) {
  return [...array].map((x) => x.toString(16).padStart(2, "0")).join("\n");
}

export function benchCodec<T>(name: string, codec: Codec<T>, value: T) {
  const encoded = codec.encode(value);
  Deno.bench(`- ${name} (${encoded.length}B) [encode] `, () => {
    codec.encode(value);
  });
  Deno.bench(`  ${name} (${encoded.length}B) [decode]`, () => {
    codec.decode(encoded);
  });
}
