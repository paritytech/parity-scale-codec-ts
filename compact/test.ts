import * as $ from "../mod.ts";
import { testCodec } from "../test-util.ts";

testCodec("compact", $.compact, [
  0,
  2 ** 8 - 1,
  2 ** 16 - 1,
  2 ** 32 - 1,
  2n ** 64n - 1n,
  2n ** 128n - 1n,
]);
