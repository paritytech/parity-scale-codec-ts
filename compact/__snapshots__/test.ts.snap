export const snapshot = {};

snapshot[`compact 0 1`] = `00`;

snapshot[`compact 255 1`] = `
fd
03
`;

snapshot[`compact 65535 1`] = `
fe
ff
03
00
`;

snapshot[`compact 4294967295 1`] = `
03
ff
ff
ff
ff
`;

snapshot[`compact 18446744073709551615n 1`] = `
13
ff
ff
ff
ff
ff
ff
ff
ff
`;

snapshot[`compact 340282366920938463463374607431768211455n 1`] = `
33
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
`;
